src/lms-ds-loader/edfi_lms_ds_loader/mssql_lms_operations.py

```diff
@@ -589,6 +633,219 @@ def soft_delete_from_production(self, table: str, source_system: str) -> None:
+        self, table: str, source_system: str
+    ) -> None:
+        """
+        Updates production records that do not have a match in the staging table
+        by setting their `deletedat` value to the current timestamp, but
+        only for the related users in the staging table.
+
+        Parameters
+        ----------
+        table: str
+            Name of the table to soft delete on, not including the `stg_`.
+        source_system: str
+            The SourceSystem currently being processed.
+        """
+
+        assert table.strip() != "", "Argument `table` cannot be whitespace"
+
+        statement = f"""
+UPDATE
+    t
+SET
+    t.DeletedAt = getdate()
+FROM
+    lms.{table} as t
+WHERE
+    t.LMSUserIdentifier IN (
+        SELECT
+            u.LMSUserIdentifier
+        FROM
+           lms.LMSUser as u
+        INNER JOIN
+            lms.stg_{table} as stg
+        ON
+            stg.LMSUserSourceSystemIdentifier = u.SourceSystemIdentifier
+        AND
+            stg.SourceSystem = u.SourceSystem
+    )
+AND
+    NOT EXISTS (
+        SELECT
+            1
+        FROM
+            lms.stg_{table} as stg
+        WHERE
+            t.SourceSystemIdentifier = stg.SourceSystemIdentifier
+        AND
+            t.SourceSystem = stg.SourceSystem
+    )
+AND
+    t.DeletedAt IS NULL
+AND
+    t.SourceSystem = '{source_system}'
+"""
+
+        row_count = self._exec(statement)
+        logger.debug(f"Soft-deleted {row_count} records in table `{table}`")
+
+    def soft_delete_from_production_for_section_and_user_relation(
+        self, table: str, source_system: str
+    ) -> None:
+        """
+        Updates production records that do not have a match in the staging table
+        by setting their `deletedat` value to the current timestamp, but
+        only for the related sections and users in the staging table.
+
+        Parameters
+        ----------
+        table: str
+            Name of the table to soft delete on, not including the `stg_`.
+        source_system: str
+            The SourceSystem currently being processed.
+        """
+
+        assert table.strip() != "", "Argument `table` cannot be whitespace"
+
+        statement = f"""
+UPDATE
+    t
+SET
+    t.DeletedAt = getdate()
+FROM
+    lms.{table} as t
+WHERE
+    t.LMSSectionIdentifier IN (
+        SELECT
+            s.LMSSectionIdentifier
+        FROM
+           lms.LMSSection as s
+        INNER JOIN
+            lms.stg_{table} as stg
+        ON
+            stg.LMSSectionSourceSystemIdentifier = s.SourceSystemIdentifier
+        AND
+            stg.SourceSystem = s.SourceSystem
+    )
+AND
+    t.LMSUserIdentifier IN (
+        SELECT
+            u.LMSUserIdentifier
+        FROM
+           lms.LMSUser as u
+        INNER JOIN
+            lms.stg_{table} as stg
+        ON
+            stg.LMSUserSourceSystemIdentifier = u.SourceSystemIdentifier
+        AND
+            stg.SourceSystem = u.SourceSystem
+    )
```

Comment:
This clause doesn't seem to be appropriate for Attendance Events.

* Assume school is tracking negative attendance
* One day a teacher accidentally marks a student as absent
* --> there's a record for the student in the file
* Next day fixes it
* --> there's no record the student in the file

This clause will ignore the missing record, since there is no replacement record for that student. I've pushed a new commit with updated sample-out data that includes this scenario.

Reply:
I think this is also true for Section Association and Section Activities. Ultimately I think the soft delete test just needs to ensure that the discovered records belong to the section that was just uploaded, without needing to look at the user identifier at all.

Reply:
Yes, I think you are right.  It's only important at the file partition level (section).